File: Design/Product of the Last K Numbers/solution_segmentTree.go
package main

type SegmentTree struct {
	Start   int
	End     int
	Product int
	Left    *SegmentTree
	Right   *SegmentTree
}

func (this *SegmentTree) Query(s, e int) int {
	if e < this.Start || s > this.End {
		return -1
	}

	if e >= this.End && s <= this.Start {
		return this.Product
	}

	this.normalize()
	leftQuery := this.Left.Query(s, e)
	rightQuery := this.Right.Query(s, e)
	if leftQuery == -1 {
		return rightQuery
	}
	if rightQuery == -1 {
		return leftQuery
	}
	return leftQuery * rightQuery
}

func (this *SegmentTree) Update(s, e, product int) {
	if e < this.Start || s > this.End {
		return
	}

	if e >= this.End && s <= this.Start {
		this.Product *= product
		return
	}

	this.normalize()
	this.Left.Update(s, e, product)
	this.Right.Update(s, e, product)

	this.Product = this.Left.Product * this.Right.Product
}

func (this *SegmentTree) normalize() {
	if this.Start != this.End {
		if this.Left == nil || this.Right == nil {
			middle := this.Start + (this.End-this.Start)/2
			this.Left = &SegmentTree{Start: this.Start, End: middle, Product: this.Product}
			this.Right = &SegmentTree{Start: middle + 1, End: this.End, Product: this.Product}
		}
	}
}

type ProductOfNumbers struct {
	root  *SegmentTree
	index int
}

func Constructor() ProductOfNumbers {
	return ProductOfNumbers{root: &SegmentTree{Start: 1, End: 40000, Product: 1}, index: 0}
}

func (this *ProductOfNumbers) Add(num int) {
	this.index++
	this.root.Update(this.index, this.index, num)
}

func (this *ProductOfNumbers) GetProduct(k int) int {
	return this.root.Query(this.index-k+1, this.index)
}

File: Sliding Window/*Longest Continuous Subarray With Absolute Diff Less Than or Equal to Limit/solution_segmentTree.ts
function longestSubarray(nums: number[], limit: number): number {
    let start = 0
    let result = 0
    let minTree = new SegmentTree(0, nums.length - 1, Math.min)
    let maxTree = new SegmentTree(0, nums.length - 1, Math.max)
    for(let [idx, num] of nums.entries()) {
        minTree.update(idx, num)
        maxTree.update(idx, num)
    }
    for(let end = 0; end < nums.length; end++) {
        let min = minTree.query(start, end)
        let max = maxTree.query(start, end)
        let diff = max - min
        while(diff > limit && start <= end) {
            start++
            min = minTree.query(start, end)
            max = maxTree.query(start, end)
            diff = max - min
        }
        if(start > end) continue
        result = Math.max(result, end - start + 1)
    }
    return result
};

class SegmentTree {
    left
    right
    start
    end
    val
    fn
    
    constructor(start, end, fn) {
        this.start = start
        this.end = end
        this.fn = fn
        if(this.fn === Math.max) {
            this.val = Number.MIN_SAFE_INTEGER
        } else {
            this.val = Number.MAX_SAFE_INTEGER
        }
    }
    
    query(start, end) {
        if(start > this.end || end < this.start) {
            if(this.fn === Math.max) {
                return Number.MIN_SAFE_INTEGER
            } else {
                return Number.MAX_SAFE_INTEGER
            }
        }
        if(this.start >= start && end >= this.end) {
            return this.val
        }
        this.down()
        let left = this.left.query(start, end)
        let right = this.right.query(start, end)
        return this.fn(left, right)
    }
    
    update(position, val) {
        if(position > this.end || position < this.start) {
            return
        }
        if(this.start === this.end) {
            this.val = val
            return
        }
        this.down()
        this.left.update(position, val)
        this.right.update(position, val)
        this.val = this.fn(this.left.val, this.right.val)
        
    }
    
    down() {
        if(this.start !== this.end) {
            if(!this.left || !this.right) {
                const middle = this.start + Math.floor((this.end - this.start) / 2)
                this.left = new SegmentTree(this.start, middle, this.fn)
                this.right = new SegmentTree(middle + 1, this.end, this.fn)
            }
        }
    }
}